extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -301,4 +348,114 @@ private void removeCookie(RoutingContext context, TenantConfigContext configCont
+            JWT jwt = auth.getJWT();
+            JsonObject tokenJson;
+
+            try {
+                tokenJson = jwt.decode(idToken);
+            } catch (Exception cause) {
+                throw new AuthenticationCompletionException(cause);
+            }
+
+            Long iat = tokenJson.getLong("iat");
+            long now = System.currentTimeMillis() / 1000;
+            Integer timeout = config.token.refreshTimeout.get();
+
+            if (now - iat >= timeout) {
+                String sub = tokenJson.getString("sub");
+
+                TokenEntry entry = refreshingTokens.computeIfAbsent(sub,
+                        new Function<String, TokenEntry>() {
+                            @Override
+                            public TokenEntry apply(String token) {
+                                return new TokenEntry(refreshToken);
+                            }
+                        });
+
+                // synchronize on the entry associated with the current subject
+                synchronized (entry) {
+                    Uni<SecurityIdentity> cf = Uni.createFrom().emitter(new Consumer<UniEmitter<? super SecurityIdentity>>() {
+                        @Override
+                        public void accept(UniEmitter<? super SecurityIdentity> emitter) {
+                            OAuth2TokenImpl token = new OAuth2TokenImpl(configContext.auth, new JsonObject());
+
+                            // always get the last token
+                            token.principal().put("refresh_token", entry.getToken());
+
+                            token.refresh(new Handler<AsyncResult<Void>>() {
+                                @Override
+                                public void handle(AsyncResult<Void> result) {
+                                    if (result.succeeded()) {
+                                        authenticate(identityProviderManager,
+                                                new IdTokenCredential(token.opaqueIdToken(), context))
+                                                        .subscribe().with(new Consumer<SecurityIdentity>() {
+                                                            @Override
+                                                            public void accept(SecurityIdentity identity) {
+                                                                // after a successful refresh, rebuild the identity and update the cookie 
+                                                                processSuccessfulAuthentication(context, configContext, token,
+                                                                        identity);
+                                                                // update the token so that blocking threads get the latest one
+                                                                entry.setToken(token.opaqueRefreshToken());
+                                                                emitter.complete(
```

Comment:
So yeah, what happens here to the blocking threads, they get the refresh token and proceed with their own refresh ? Or can we just let those blocking threads pick up the refreshed token/etc and continue without the refresh ?

Reply:
Yes. Thing is that we would need to know if a refresh happened before to make the second point work. I think is easier to just let them refresh and get a new set of tokens.

Reply:
So can this bit perform blocking operations? If so you need to check if you are on the IO thread and dispatch to an executor.

Reply:
Everything is within a call to `authenticate`  which will end up executing as blocking (by the identity provider).

And here is the place I mentioned to you about the Vert.x OAuth2 always execute in the event loop thread. If you check the thread before the `authenticate` call you'll see it runs in the event loop thread. But if you check the thread within the block you are commenting you'll see it does run in an executor thread.

Reply:
@pedroigor OK, so we just keep recycling for every concurrent request. OK, I'm still a bit unclear about where these concurrent requests can originate from with the same `sub`, I'll ping you later to understand better :-) but obviously there are all sort of variations possible with 'sub' being reused.
I'll open an issue to track the even loop issue